File: python/sparktk/frame/constructors/import_tensorflow.py
# vim: set encoding=utf-8

#  Copyright (c) 2016 Intel Corporation 
#
#  Licensed under the Apache License, Version 2.0 (the "License");
#  you may not use this file except in compliance with the License.
#  You may obtain a copy of the License at
#
#       http://www.apache.org/licenses/LICENSE-2.0
#
#  Unless required by applicable law or agreed to in writing, software
#  distributed under the License is distributed on an "AS IS" BASIS,
#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
#  See the License for the specific language governing permissions and
#  limitations under the License.
#

from sparktk import TkContext


def import_tensorflow(tf_path, schema=None, tc=TkContext.implicit):
    """
    Create a frame with data from a TensorFlow records file

    TensorFlow records are the standard data format for TensorFlow. The recommended format for TensorFlow is a TFRecords file
    containing tf.train.Example protocol buffers. The tf.train.Example protocol buffers encodes (which contain Features as a field).
    https://www.tensorflow.org/how_tos/reading_data

    During Import, API parses TensorFlow DataTypes as below

    * Int64List => IntegerType or LongType
    * FloatList => FloatType or DoubleType
    * Any other DataType (Ex: String) => BytesList

    Parameters
    ----------

    :param tf_path:(str) Full path to TensorFlow records
    :param schema: (Optional(list[tuple(str, type)] or list[str])) The are different options for specifying a schema:

    * Provide the full schema for the frame as a list of tuples (string column name and data type)
    * Provide the column names as a list of strings.  Column data types will be inferred, based on the data.

    :return: a frame
    """

    if schema is not None:
        scala_frame_schema = tc.jutils.convert.to_scala_frame_schema(schema)
    else:
        scala_frame_schema = schema

    scala_frame = tc.sc._jvm.org.trustedanalytics.sparktk.frame.internal.constructors.ImportTensorflow.importTensorflow(tc._scala_sc,
                                                                                                                        tf_path,
                                                                                                                        tc.jutils.convert.to_scala_option(scala_frame_schema))

    from sparktk.frame.frame import Frame
    return Frame(tc, scala_frame)

File: sparktk-core/src/main/scala/org/trustedanalytics/sparktk/frame/internal/constructors/ImportTensorflow.scala
/**
 *  Copyright (c) 2016 Intel Corporation 
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */
package org.trustedanalytics.sparktk.frame.internal.constructors

import org.apache.commons.lang.StringUtils
import org.apache.hadoop.io.{ BytesWritable, NullWritable }
import org.apache.spark.SparkContext
import org.tensorflow.example.Example
import org.tensorflow.hadoop.io.TFRecordFileInputFormat
import org.trustedanalytics.sparktk.frame.{ Frame, FrameSchema, TensorflowInferSchema }
import org.trustedanalytics.sparktk.frame.internal.serde.DefaultTfRecordRowDecoder

object ImportTensorflow {
  /**
   * Creates a frame using TensorFlow Records path with specified schema
    *
   * TensorFlow records are the standard data format for TensorFlow. The recommended format for TensorFlow is a TFRecords file
    * containing tf.train.Example protocol buffers. The tf.train.Example protocol buffers encodes (which contain Features as a field).
    * https://www.tensorflow.org/how_tos/reading_data
    *
   * During Import, API parses TensorFlow DataTypes as below
   *
   * Int64List => IntegerType or LongType
   * FloatList => FloatType or DoubleType
   * Any other DataType (Ex: String) => BytesList
   *
   * @param sc sparkcontext
   * @param sourceTfRecordsPath Full path to TensorFlow records on HDFS/Local filesystem
   * @param schema Optional frame schema to use during import. If not defined, then the schema is inferred from the TensorFlow records
   * @return frame with data from TensorFlow records
   */
  def importTensorflow(sc: SparkContext, sourceTfRecordsPath: String, schema: Option[FrameSchema] = None): Frame = {
    require(StringUtils.isNotEmpty(sourceTfRecordsPath), "path should not be null or empty.")

    val rdd = sc.newAPIHadoopFile(sourceTfRecordsPath, classOf[TFRecordFileInputFormat], classOf[BytesWritable], classOf[NullWritable])

    val exampleRdd = rdd.map {
      case (bytesWritable, nullWritable) => Example.parseFrom(bytesWritable.getBytes)
    }


    var finalSchema = if (schema.isEmpty) Some(TensorflowInferSchema(exampleRdd)).get else schema.get

    val resultRdd = exampleRdd.map(example => DefaultTfRecordRowDecoder.decodeTfRecord(example, finalSchema))
    new Frame(resultRdd, finalSchema)
  }

}
